File: src/config/env.ts
import { z } from "zod/v4";

const envSchema = z.object({
	PORT: z.coerce.number().default(3000),
	NODE_ENV: z
		.enum(["development", "production", "test"])
		.default("development"),

	// Database
	MONGODB_URI: z.string().default("mongodb://localhost:27017/dict"),
	REDIS_URI: z.string().default("redis://localhost:6379"),

	// Auth
	JWT_SECRET: z.string().min(1, "JWT_SECRET environment variable is required"),

	// OpenTelemetry
	OTEL_EXPORTER_OTLP_ENDPOINT: z
		.string()
		.default("http://localhost:4318/v1/traces"),

	// Rate Limiting
	RATE_LIMIT_BUCKET_SIZE: z.coerce.number().default(60),
	RATE_LIMIT_REFILL_SECONDS: z.coerce.number().default(60),
});

// Validate process.env
// We accept partial here because we might want to override defaults but if JWT_SECRET is missing it should fail
const _env = envSchema.safeParse(process.env);

if (!_env.success) {
	console.error("❌ Invalid environment variables:");
	console.error(JSON.stringify(_env.error.format(), null, 2));
	process.exit(1);
}

export const env = _env.data;

File: go/internal/config/env.go
package config

import (
	"log"
	"os"
	"strconv"
)

type Config struct {
	Port                   int
	Environment            string
	MongoDBURI             string
	RedisURI               string
	JWTSecret              string
	OTELExporterEndpoint   string
	RateLimitEnabled       bool
	RateLimitBucketSize    int
	RateLimitRefillSeconds int
}

var Env *Config

func Load() {
	port, _ := strconv.Atoi(getEnvOrDefault("PORT", "3000"))
	rateLimitEnabled := getEnvOrDefault("RATE_LIMIT_ENABLED", "true")
	rateLimitBucketSize, _ := strconv.Atoi(getEnvOrDefault("RATE_LIMIT_BUCKET_SIZE", "60"))
	rateLimitRefillSeconds, _ := strconv.Atoi(getEnvOrDefault("RATE_LIMIT_REFILL_SECONDS", "60"))

	jwtSecret := os.Getenv("JWT_SECRET")
	if jwtSecret == "" {
		log.Fatal("JWT_SECRET environment variable is required")
	}

	Env = &Config{
		Port:                   port,
		Environment:            getEnvOrDefault("GO_ENV", "development"),
		MongoDBURI:             getEnvOrDefault("MONGODB_URI", "mongodb://localhost:27017/dict"),
		RedisURI:               getEnvOrDefault("REDIS_URI", "redis://localhost:6379"),
		JWTSecret:              jwtSecret,
		OTELExporterEndpoint:   getEnvOrDefault("OTEL_EXPORTER_OTLP_ENDPOINT", "http://localhost:4318/v1/traces"),
		RateLimitEnabled:       rateLimitEnabled != "false" && rateLimitEnabled != "0",
		RateLimitBucketSize:    rateLimitBucketSize,
		RateLimitRefillSeconds: rateLimitRefillSeconds,
	}
}

func getEnvOrDefault(key, defaultValue string) string {
	if value := os.Getenv(key); value != "" {
		return value
	}
	return defaultValue
}
